### ISSUE TITLE
New viewport units don't work on iOS

### ISSUE THREAD

**Issue by emilio** [mozilla-mobile/firefox-ios#11574]
They work on Safari and Chrome on iOS. There are new APIs that should be used to make this work:

* https://developer.apple.com/documentation/webkit/wkwebview/3974127-setminimumviewportinset?language=objc
* https://github.com/WebKit/WebKit/blob/3197e7a821f6af5a0f5ab9ac654718668d7a82e4/Source/WebKit/UIProcess/API/Cocoa/WKWebView.h#L628

┆Issue is synchronized with this [Jira Spike](https://mozilla-hub.atlassian.net/browse/FXIOS-4747)

**Comment by data-sync-user** [comment]
➤ Razvan Litianu commented:

Hi Winnie Teichmann, can you please give an example where this happens? And possibly a counter example from Chrome/Safari?

**Comment by dholbert** [comment]
It looks like https://bokand.github.io/demo/urlbarsize.html is one demo that shows the behavior difference.  Try viewing that page in Firefox vs. Safari/Chrome on iOS.

In Firefox, the colorful bars at the bottom-right are all the same height as each other (initially as well as after you scroll).
In Safari/Chrome, the rightmost two blue bars and the rightmost two brown bars are taller than the others at initial load; and various bars get taller when you scroll (as the dynamic toolbar goes offscreen and the viewport gets larger).

**Comment by bramus** [comment]
@dholbert FYI: There’s a bunch of demos available on https://interop-2022-viewport.netlify.app/ for you to test all these, as part of last year’s [Interop2022 Viewport Investigation](https://github.com/web-platform-tests/interop-2022-viewport).

More specifically https://interop-2022-viewport.netlify.app/combined/viewport-units/ – which is an updated version of bokan’s demo – is relevant here.

Related: Firefox _(both on iOS and Android)_ resize the ICB when the Virtual Keyboard gets shown. Chrome on Android also used to do that, [but no longer does as of version 108](https://developer.chrome.com/blog/viewport-resize-behavior/#changing-the-default-behavior-in-chrome-108) and now lines up with Safari on iOS. This ICB resizing might need to be tackled first, as the Viewport Units are derived from them.

**Comment by jteppinette** [comment]
Related, when opening a new tab and the bottom search bar has been minimized (by scrolling down), the new tab will be created with incorrect viewport height, until a scroll event occurs.

This impacts anything positioned relative to the viewport. i.e. items positioned at bottom of the viewport will be below the search bar and invisible.

**Comment by data-sync-user** [comment]
➤ Laurie Marceau commented:

Turning this ticket into a spike to document finding and path forward to implement the view ports within our app.

**Comment by dyviknilsen** [comment]
This issue still persists; svh and lvh act like dvh, which is unfortunate as more and more websites are adopting dynamic viewport units e.g. to position and trigger scroll-based animations. These still break in Firefox.

**Comment by PullRequestTimeout** [comment]
It's been 789 days. The horrors persist, but so do I.

**Comment by atsakiridis** [comment]
Any updates here, this is still happening with latest firefox on iOS

**Comment by acki** [comment]
+1

**Comment by jharrje** [comment]
This issue has caused me great mental torment :)

**Comment by Cap0n3** [comment]
Yep, same here... 100svh is behaving the same way as 100dvh on Firefox for iOS. WTF.

**Comment by wraybowling** [comment]
This seems to have passed some sort of verification through #870, and a couple of flashy demos including a new one linked in #22607, but all I can say is I wasted another day thinking my sizing was broken only to realize Firefox on iOS doesn't understand `lvh`.

**Comment by regiside** [comment]
Adding my name to the list of web develops frustrated by not being able to use svh in firefox without a super janky experience for my users. If anyone could implement a solution to stop svh and lvh from being attached to the visual viewport I'd be so grateful! Or at least just remove support for these so we can use @supports to set up work arounds.

**Comment by kieferslaton** [comment]
Has anyone found a way around this that mimics a consistent svh? I know I could do lvh by getting the overall device screen height, but I really need my elements to stay at a fixed height that is 100% of the viewport when the browser bars are in place.

**Comment by regiside** [comment]
Hey @kieferslaton, my solution was to just run a bit of javascript to get the viewport height on page load and set it as a css variable:
`window.addEventListener('load', function() {
  document.documentElement.style.setProperty('--small-viewport-height', document.documentElement.clientHeight + 'px');
});`

You can then set your element to 100% viewport height like so:
```
:root{--small-viewport-height:100vh; /*this is replaced by javascript on page load.*/}

.full-height-section {
     min-height: var(--small-viewport-height);
}
```

**Comment by bramus** [comment]
I believe using the following is more robust than relying on `window.innerHeight` (which can change).

```js
document.documentElement.clientWidth;
document.documentElement.clientHeight;
```

See https://github.com/web-platform-tests/interop-2022-viewport/blob/main/explainers/icb.md for a reference.

**Comment by valler** [comment]
@regiside @bramus

Page load is independent from viewport size. Nothing stops you from visiting a page with a hidden toolbar on iOS, unless FF doesn't have that feature and I was mistakenly thinking it would. What I mean by that is, there's browsers which allow toggling the viewport explicitly, rather than only via scroll. And unless your script is able to sniff out FFM, you're breaking the experience for users of browsers with that feature. This script would then confuse `lvh` for `svh`. If you also happen to use resize events to then toggle between styles, you've basically broken your layout for everyone switching back to an actual small viewport. Also BFC (back-forward-cache) can toggle between hidden and shown toolbars without firing load, so this script won't cover that. But again, maybe this is browser specific and FF was not one of the browsers behaving this way. Either way this script would probably break your layout for a subset of users.

@kieferslaton simply put: there is none. Brave has or has had the same problem (their code is based on FF) and a brave contributor mentioned on an issue I opened, that the implementation only has one notion of viewport. So basically all you have is a `vh` that behaves like `dvh` and all the other values are also just the same. And there's no way of sniffing out wether your current view corresponds to small or large (initially), because in general a page can be opened in either state; only on resize you might be able to figure out if its shrinking or growing in a way that indicates a mobile `svh`/`lvh` switch. But anything relying on having both sizes available (at any time) for styling is not doable.

**Comment by regiside** [comment]
Hey @valler, I've just tested Safari, Chrome & Firefox on iOS and each browser (because they're all webkit) loads every new page with the browser bars showing, even if they were hidden after scroll on the previous page. Do you know of a specific setting in any iOS browser that changes that?

Test here: [https://cogentstudios.com/browser-height-test/]

The one major down side of this approach is that it will not work to update this when the user resizes their browser because Firefox is telling webkit to resize the viewport when the user scrolls so if you run the javascript on window.resize it will run on scroll on Firefox causing janky sizing behavior. You could work around this by only updating the height on resize above a certain width, or by trying to detect if it’s a mobile browser in general.

This seems to be the best and somewhat usable javascript solution for the average user. Hopefully between Webkit API updates and Firefox developers we can eventually get to a built in unit that behaves consistently across all browsers.

P.S. if you're using document.documentElement.clientHeight make sure and properly declare your document type `<!DOCTYPE html>` otherwise you'll get some whacky numbers on high dpi screens.

**Comment by kieferslaton** [comment]
@regiside My experience has been consistent with what @valler is saying, which may or may not be due to the fact that I am using custom page transitions with Nuxt, so there is no distinct reload event in the browser. Regardless, none of the solutions involving the client height work for me because there are inconsistencies with whether the DOM is aware of the browser bar at the time the transition happens. Fortunately, in my particular application it isn't the end of the world if these elements don't completely fill the screen height, so I'm just setting them to 85% of the overall screen height excluding the browser bar and hoping that's enough clearance across most browsers to show the whole content while the browser bar is also showing.

**Comment by bramus** [comment]
Dixit @valler in https://github.com/mozilla-mobile/firefox-ios/issues/11574#issuecomment-3433456170:

> Either way this script would probably break your layout for a subset of users.

If you are talking about [my snippet](https://github.com/mozilla-mobile/firefox-ios/issues/11574#issuecomment-3428365300) that uses `document.documentElement.clientWidth` and `document.documentElement.clientHeight`, then you are mistaken. These properties (in non-quirks mode) return the size of the ICB, which matches up with the size of the small viewport.

See https://github.com/web-platform-tests/interop-2022-viewport/blob/main/explainers/icb.md#measuring-the-icb for details and references to why this works.

It’s when using `window.innerHeight` (the snippet I was replying to) that you can run into the issues which you have detailed. You’d then also run into issues in Safari, as that is the only browser that resizes the `innerHeight`/`innerWidth` on pinch-zoom.

**Comment by regiside** [comment]
@kieferslaton I see, but if you set the variable when the site first loads it should always give you the small viewport height on webkit browsers. You wouldn't want to reset that variable per page transition because then it could give you a value without the browser bars. (Just tested and seeing same behavior on chrome and Firefox for android where the address bar is shown every time a new site is loaded.)

**Comment by valler** [comment]
@bramus and the rest of course. what I mean is that browsers like Safari allow you to explicitly hide the toolbar from the browser menu and this setting is typically more persistent than the one that hides the toolbar when you scroll down. I can't remember and haven't had the time to test, which type of navigation brings back the toolbar (if any), or if it requires a fresh page reveal/show/load. Also Safari fires resize events when toggling the viewport with this setting, which it doesn't when reacting to scrolls. All together because of that, I wouldn't be so sure to be able to rely on ICB being small, because a) I'm assuming the scripts do not perform UA sniffing and b) the toolbar won't reset at least on same origin navigation (but maybe I'm not remembering this one correctly). as it stands I tend to lean towards "there's no general solution". maybe for some special case scenarios one can find cross browser workarounds (e.g. SPAs hosted at the origin which don't change the URL on navigation). In summary: If your origin is `example.com` with a landing page, and your script only applies to some other page or app on `example.com/myapp`, then I'd rather double check if navigating from root to app with an explicitly hidden toolbar works as expected on all target browsers.

**Comment by valler** [comment]
@regiside yes, please test on iOS Safari with toolbars hidden via the browser menu. reaching this settings/switch is different for different versions of iOS Safari. It's either reachable via `aA` on older versions, or via the page menu icon to the left of the url/search bar > then the ellipsis on the bottom right `...`. > then in both variants the option you're looking for is written as "Hide Toolbar".

**Comment by valler** [comment]
Having thought about this again, even if the "Hide Toolbar" option turns out to not survive any form of navigation, the problem of page load events remains. Page show/hide are the only reliable candidates for earliest/latest life cycle events for setting up and tearing down app state. load does not fire in all scenarios. so basically sniffing (small) ICB can only happen 
 
- with a FOUC (or initial/delayed layout shift, [or whatever you're doing with `svh` if it's not layout])
- or only for a subset of page show events (when sticking to the load event, ignoring its unreliability)
- when you need it only much later anyway (but that's tricky, unless you can guarantee where exactly one lands when returning to a page, or you have some form of interaction not tied to a location/coordiante to rely on)

I'd never opt into the first two just to polyfill a minute CSS feature, killing the experience of the majority of iOS users.

Edit: Another option would be to use a sync inline script. This could actually run before rendering happens, with maybe not too much of a penalty. But the last time I've used one of these is so long ago, I don't even know how they behave with BFC out of the top of my head.

**Comment by regiside** [comment]
Hey @valler, Thanks to css variables there is little to no layout shift. We can initially set the --small-viewport-height css variable using the browsers svh unit.
`:root{--small-viewport-height:100svh;}`
Then we just replace that value using javascript to a static pixel value to avoid the issue in firefox on iOS where divs using svh [jump in size when the user scrolls](https://github.com/mozilla-mobile/firefox-ios/issues/26271). 

(I did test hiding the address bar in safari, it works fine if someone has hidden the bar and then clicks on a link to your page. The setting is not permanent, as soon as you go to navigate to a different page, the address bar comes back. If they're on your page and then hide the address bar, the viewport will be larger than --small-viewport-height variable. But in my opinion it's a small enough use case where I'd prefer a better experience for the vast majority.)

**Comment by janbrasna** [comment]
I've been looking into the various WebKit implementations, mostly from reports mentioning that Chrome iOS behaves correctly, but I'm unable to confirm the expected/correct behavior in any browser, incl. Chrome iOS, Edge iOS, Brave iOS, Vivaldi iOS, Orion iOS et al.: every single WKWebKit browser changes the unit length when scrolling to some extent.

FxiOS v146:

> <img width="1977" height="1488" alt="Firefox" src="https://github.com/user-attachments/assets/d185c995-47aa-4a50-a1d4-ffa20e225ad6" />

(this does only "one" single transition hiding the toolbar)

CriOS v144:

> <img width="1134" height="744" alt="Chrome" src="https://github.com/user-attachments/assets/dae0da67-ed4d-4414-996d-44be736f736f" />

(here the many events come from easing the toolbar out and back in, resizing the viewport many times under the thumb; also NB the final dimensions are slightly different than the initial metrics in the same state…)

EdgiOS v144:

> <img width="1920" height="856" alt="Edge" src="https://github.com/user-attachments/assets/bd90122c-d467-4eb6-8307-27cc88b5468c" />

(here the only difference is even if flicked rather slowly, the transition is animated fast enough to only cause just a handful of resizes, compared to Chrome's dozens changes, feeling less jittery…)

- - - -

The only correct readout seems to be currently in closed-source system default Safari, not having to follow any app store rules or rely on public APIs:

Safari v26:

> <img width="1987" height="1488" alt="Safari" src="https://github.com/user-attachments/assets/8424f272-ffba-4343-82bd-94c6a144e57e" />

Interestingly, even Safari does one initial resize.

**Comment by janbrasna** [comment]
Love that there are passionate folks in the community who actually tried the APIs on their own to see the impact — quoting e.g. from [Chrome and Firefox don't support viewport units on iOS](https://studiowaslo.com/essays/broken-ios-viewport-units/):

> "As far as I can tell, the fault lies with Google and Mozilla. Two years ago, around the time Apple shipped viewport units for Webkit, they added the minimumViewportInset and maximumViewportInset APIs to Webkit so that third-party browser developers could support these new units. These APIs are dead simple—all you have to do is set the size of the viewport at its smallest and at its largest and this enables Webkit to properly define the small and large viewport for units like svh. However, neither Google nor Mozilla added the **two lines of code** needed to make these units work."

> "Edit: I attempted to fix this issue for Firefox on my own since it’s open source, and it **seems much harder** than I initially thought."

> "Obviously, the problem isn’t fixed. […] This means that fixing this issue requires a bespoke implementation of the viewport units or a rewrite of the various UI components involved in this […] Lesson learned: **things are actually much more complicated** than I think and I shouldn’t be such a naive and confident fool >__<." ;D

_(emphasis mine)_
